yq/toolchain/toolchain.bzl

```diff
@@ -0,0 +1,134 @@
+    attrs = {
+        "bin": attr.label(
+            mandatory = True,
+            allow_single_file = True,
+        ),
+    },
+)
+
+def _yq_toolchains_repo_impl(rctx):
+    # Expose a concrete toolchain which is the result of Bazel resolving the toolchain
+    # for the execution or target platform.
+    # Workaround for https://github.com/bazelbuild/bazel/issues/14009
+    starlark_content = """# @generated by @yq.bzl//yq/toolchain:toolchain.bzl
+
+# Forward all the providers
+def _resolved_toolchain_impl(ctx):
+    toolchain_info = ctx.toolchains["@yq.bzl//yq/toolchain:type"]
+    return [
+        toolchain_info,
+        toolchain_info.default,
+        toolchain_info.yqinfo,
+        toolchain_info.template_variables,
+    ]
+
+# Copied from java_toolchain_alias
+# https://cs.opensource.google/bazel/bazel/+/master:tools/jdk/java_toolchain_alias.bzl
+resolved_toolchain = rule(
+    implementation = _resolved_toolchain_impl,
+    toolchains = ["@yq.bzl//yq/toolchain:type"],
+)
+"""
+    rctx.file("defs.bzl", starlark_content)
+
+    build_content = """# @generated by @yq.bzl//yq/toolchain:toolchain.bzl
+#
+# These can be registered in the workspace file or passed to --extra_toolchains flag.
+# By default all these toolchains are registered by the yq_register_toolchains macro
+# so you don't normally need to interact with these targets.
+
+load(":defs.bzl", "resolved_toolchain")
+
+resolved_toolchain(name = "resolved_toolchain", visibility = ["//visibility:public"])
+
+"""
+
+    for [platform, meta] in YQ_PLATFORMS.items():
+        build_content += """
+toolchain(
+    name = "{platform}_toolchain",
+    exec_compatible_with = {compatible_with},
+    toolchain = "@{user_repository_name}_{platform}//:yq_toolchain",
+    toolchain_type = "@yq.bzl//yq/toolchain:type",
+)
+""".format(
+            platform = platform,
+            user_repository_name = rctx.attr.user_repository_name,
+            compatible_with = meta.compatible_with,
+        )
+
+    # Base BUILD file for this repository
+    rctx.file("BUILD.bazel", build_content)
+
+yq_toolchains_repo = repository_rule(
+    _yq_toolchains_repo_impl,
+    doc = """Creates a repository with toolchain definitions for all known platforms
+     which can be registered or selected.""",
+    attrs = {
+        "user_repository_name": attr.string(doc = "Base name for toolchains repository"),
+    },
+)
+
+def _yq_host_alias_repo(rctx):
+    ext = ".exe" if repo_utils.is_windows(rctx) else ""
+
+    # Base BUILD file for this repository
+    rctx.file("BUILD.bazel", """# @generated by @yq.bzl//yq/toolchain:toolchain.bzl
+package(default_visibility = ["//visibility:public"])
+exports_files(["yq{ext}"])
+""".format(
+        ext = ext,
+    ))
+
+    rctx.symlink("../{name}_{platform}/yq{ext}".format(
+        name = rctx.attr.name,
+        platform = repo_utils.platform(rctx),
+        ext = ext,
+    ), "yq{ext}".format(ext = ext))
+
+yq_host_alias_repo = repository_rule(
```

Comment:
Could we drop this? The resolved target should serve pretty much the same use cases and sniffing the host is a code smell as it breaks in cross-platform builds

Reply:
yeah it was a smell that it needed the repo_utils to detect a platform as well. This wrong thing was in bazel-lib too.